Python/ceval.c

```diff
@@ -4617,6 +4635,45 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, _PyInterpreterFrame *frame, int
             DISPATCH();
         }
 
+        TARGET(PRECALL_PY_CLASS) {
+            SpecializedCacheEntry *cache = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &cache[0].adaptive;
+            _PyCallCache *cache1 = &cache[-1].call;
+            int original_oparg = cache->adaptive.original_oparg;
+            int is_method = (PEEK(original_oparg + 2) != NULL);
+            DEOPT_IF(is_method, PRECALL);
+            PyObject *cls = PEEK(original_oparg + 1);
+            DEOPT_IF(!PyType_Check(cls), PRECALL);
+            PyTypeObject *cls_t = (PyTypeObject *)cls;
+            DEOPT_IF(cls_t->tp_version_tag != cache0->version, PRECALL);
+            assert(cls_t->tp_flags & Py_TPFLAGS_HEAPTYPE);
+            PyObject *init = ((PyHeapTypeObject *)cls_t)->_spec_cache.init;
+            assert(PyFunction_Check(init));
+            DEOPT_IF(((PyFunctionObject *)(init))->func_version != cache1->func_version, PRECALL);
+            DEOPT_IF(cls_t->tp_new != PyBaseObject_Type.tp_new, PRECALL);
```

Comment:
TBH, I don't see why we need this. `__new__` can be overriden for all we care. It just needs to return *something*. I'm just playing safe for now.

Reply:
If `__new__` returns an instance of a different class, `__init__` is not called, so I do think it matters. https://docs.python.org/3.10/reference/datamodel.html#object.__new__

Reply:
Wow TIL! Thanks.